model-optimizer/extensions/front/tf/CTCGreedyDecoder.py

```diff
@@ -63,20 +78,48 @@ def replace_sub_graph(self, graph: Graph, match: dict):
 
         # update the TensorFlow infer function for the CTCGreedyDecoder to make necessary changes with the second input
         decoder_node['old_infer'] = decoder_node.infer
-        decoder_node.infer = __class__.tf_greedy_decoder_infer
+        decoder_node.infer = tf_greedy_decoder_infer
         return {}
 
+
+class CTCGreedyDecoderReplacement2(FrontReplacementSubgraph):
+    """
+    The TF implementation of the CTCGreedyDecoder produces a tuple with three tensors: indices, values and dense shape.
+    All tensors in the tuple represent SparseTensor and are inputs to SparseToDense operation.
+    This replacer matches CTCGreedyDecoder and SparseToDense operations and removes the SparseToDense, because Inference Engine 
+    implementation of the CTCGreedyDecoder produces regular tensor as output.
+
+    The second input to the CTCGreedyDecoder in the TensorFlow is a 1D tensor with sequence lengths. In the Inference
+    Engine the second input to the CTCGreedyDecoder is a 2D tensor where the first element in each row is equal to 0
+    and all others are equal to 1. The length of the row is equal to the sequence length. The replacer modifies the
+    second input to be compatible with the Inference Engine CTCGreedyDecoder layer implementation.
+    """
+    enabled = True
+
     @staticmethod
-    def tf_greedy_decoder_infer(node: Node):
-        sequence_length_node = node.in_node(1)
-        if sequence_length_node.value is None:
-            raise Error('The second input to the CTCGreedyDecoder node "{}" is not constant. This case is not '
-                        'supported with the Inference Engine.'.format(node.soft_get('name')))
-        # the batch size is the dimension with index 1 for the layer CTCGreedyDecoder
-        new_value = np.ones([node.in_node(0).shape[1], sequence_length_node.value[0]])
-        new_value[:, 0] = 0
-        new_value = np.transpose(new_value)
-        sequence_length_node.value = new_value
-        sequence_length_node.shape = int64_array(sequence_length_node.value.shape)
-
-        node.old_infer(node)
+    def pattern(**kwargs):
+        return dict(
+            nodes=[
+                ('decoder', dict(op='CTCGreedyDecoder')),
+                ('sparse_to_dense', dict(op='SparseToDense')),
+            ],
+            edges=[
+                ('decoder', 'sparse_to_dense', {'out': 0, 'in' : 0}),
+                ('decoder', 'sparse_to_dense', {'out': 2, 'in' : 1}),
+                ('decoder', 'sparse_to_dense', {'out': 1, 'in' : 2}),
+            ]
+        )
+
+    def nodes_to_remove(self, graph: Graph, match: dict):
+        return [match['sparse_to_dense']]
+
+    def replace_sub_graph(self, graph: Graph, match: dict):
+        decoder_node = match['decoder']
```

Comment:
As we have discussed in the chat we should completely rewrite these transformations code. Looks like there are a lot of legacy code which is not actual and transformation can be implemented much easier.

Reply:
Now I do avoid changes in CTCGreedyDecoder.py file. I will fix it in separate commit.